operators/src/pro/cache/cache_tiles.rs

```diff
@@ -314,23 +309,354 @@ impl<T: Pixel> Stream for CacheTileStream<T> {
                 && tile.time.intersects(&query.time_interval)
             {
                 *idx = i + 1;
-                return std::task::Poll::Ready(Some(Ok(tile.clone())));
+
+                // create a future that decodes the tile
+                let future_data = data.clone();
+                state.set(Some(crate::util::spawn_blocking(move || {
+                    let data = future_data;
+                    let idx = i;
+                    let element = RasterTile2D::<T>::from_stored_element_ref(&data[idx]);
+                    Ok(element)
+                })));
+
+                // maybe the future is already finished, so we can try to return it immediately
+                let tile_future = state
+                    .as_mut()
+                    .as_pin_mut()
+                    .expect("The future must exist since we just set it");
```

Comment:
you you check the future before setting it into the state, so you don't have to unwrap it here?

Reply:
[2df1dd8](https://github.com/geo-engine/geoengine/pull/848/commits/2df1dd88bec22458d7aacc2e33dd6eddd865b19b)
+
[59de662](https://github.com/geo-engine/geoengine/pull/848/commits/59de6629c6ee4064a1b574d91b566410a2629891) (Copilot forgot the to increase the index)

Reply:
ok but why do you need two streams now (outer and inner)?

Reply:
its not a second stream just an "inner class". when using the pin project stuff it is not that easy to call methods on self. Using this wrapper we can call "next_idx" on the inner struct.